### PR TITLE
Decouple from rails with an optional Railtie

### PR DESCRIPTION
**Pull request by thefloweringash**
This should hopefully make this easier to use outside of a rails project, and consequently easier to test since it would no longer require a rails project compatible with each rails version.

### CODE REVIEW

**Review by chrisandreae — COMMENTED** [review]
For the configuration, how do you feel about a `configure!` block like we have for [ViewModels](https://github.com/iknow/iknow_view_models/blob/master/lib/view_model/config.rb#L12) or [LoadableConfig](https://github.com/iknow/loadable_config/blob/master/lib/loadable_config.rb#L51)? I think I prefer it to independent write-once setters.

Will you update the tests to not require the Rails dependency?

**Comment by thefloweringash** [comment]
> For the configuration, how do you feel about a `configure!` block like we have for [ViewModels](https://github.com/iknow/iknow_view_models/blob/master/lib/view_model/config.rb#L12) or [LoadableConfig](https://github.com/iknow/loadable_config/blob/master/lib/loadable_config.rb#L51)? I think I prefer it to independent write-once setters.

It's a little more awkward to integrate with the initializer based defaulting. I could buffer the defaults on the railtie, and then when both logger and cache are available, send them to the library. It'd also mean either requiring the user config to be exhaustive, or making a private setter. For example, if the user says `IknowCache.configure! { logger Rails.logger }`, then the `cache` should still default to `Rails.cache` (which will become available later), but it can't be set via the single-use `configure!`.

> Will you update the tests to not require the Rails dependency?

Yes, but this change is nicely self-contained. I was going to give the tests a once-over in #6.

**Comment by chrisandreae** [comment]
I hadn't imagined anything more complex than the Railtie coming with a baked in `configure!` block, and if you're not using Rails then you bring your own. I accept that supporting mixing and matching would be harder, but do you believe it's necessary? Also, would a railtie of `configure! { blah-rails-defaults } unless configured?` be sufficient?

**Comment by thefloweringash** [comment]
> I hadn't imagined anything more complex than the Railtie coming with a baked in `configure!` block,

Specifically I want it to run after the rails initializers `:initialize_logger` and `:initialize-cache`, (see the [list of initializers for rails v5](https://github.com/rails/rails/blob/v5.2.6/guides/source/configuring.md#initializers)). So either two blocks that call some kind of `do_configure! if logger_hook_called? && cache_hook_called? && !configured?`, or attach to a later hook like `finisher_hook`

> and if you're not using Rails then you bring your own. I accept that supporting mixing and matching would be harder, but do you believe it's necessary? Also, would a railtie of `configure! { blah-rails-defaults } unless configured?` be sufficient?

I don't think I might requiring the user to be exhaustive if they're doing their own configuration.

**Comment by chrisandreae** [comment]
iirc, initializers get an `after:` keyword argument to specify their dependencies: do they need to be snuck in like that?

**Comment by thefloweringash** [comment]
Moved to a `config.after_initialize`, which is not all that precise, but very simple. How does it look to you now?

File: lib/iknow_cache.rb [hunk 1]
```diff
@@ -1,6 +1,42 @@
...
+    @configured
+  end
+
+  def self.configure!(&block)
+    raise ArgumentError.new("Already configured!") if configured?
+
+    conf = Config.new
+    ConfigWriter.new(conf).instance_eval(&block)
+
+    @logger = conf.logger
```
Comment (chrisandreae) [inline lib/iknow_cache.rb]: Should there be an exposed `IknowCache.config`, that `.logger` and `.cache` delegate to?

Comment (thefloweringash) [reply lib/iknow_cache.rb]: I considered it but didn't get around to it. You're right that it is nicer though. Added.

**Review by chrisandreae — APPROVED** [review]

File: lib/iknow_cache.rb [hunk 2]
```diff
@@ -1,5 +1,8 @@
 # frozen_string_literal: true
 
+require "active_support"
+require "active_support/core_ext/module/delegation"
```
Comment (chrisandreae) [inline lib/iknow_cache.rb]: Given that we no longer rely on `Rails.logger` and `Rails.cache` but only things with similar interfaces, and configure that up front, do we want/need this dependency outside tests at all?

Comment (thefloweringash) [reply lib/iknow_cache.rb]: The project is unlikely to be used outside of rails, but `delegate` is such a trivial dependency that it's probably worth removing. 👍

**Pull request merged**